### ISSUE TITLE
Differentiate between partially unknown variable types and "fully" unknown variable types

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#3066]
**Is your feature request related to a problem? Please describe.**
In working on the pandas stubs and the tests we are creating, I'd like to differentiate between cases where a variable type is partially unknown versus fully unknown.  Here's an example, using pyright 1.1.222, python 3.9, and numpy 1.22.2:

```python
import numpy as np

foo: np.ndarray = np.array([1, 2, 3])


class Foo:
    def __init__(self, x):
        self.x = x

    @property
    def x(self):
        return self.x

    @x.setter
    def x(self, x):
        self.x = x


goo = Foo(10).x
reveal_type(goo)
```
Using the setting file `partial.json` with the contents:
```json
{
"reportUnknownVariableType": true
}
```
I get the following report:
```
  c:\Code\Misc\pytypes\partial.py:3:1 - error: Type of "foo" is partially unknown
    Type of "foo" is "ndarray[Unknown, dtype[Any]]" (reportUnknownVariableType)
  c:\Code\Misc\pytypes\partial.py:12:16 - error: Return type is unknown (reportUnknownVariableType)
  c:\Code\Misc\pytypes\partial.py:19:1 - error: Type of "goo" is unknown (reportUnknownVariableType)
  c:\Code\Misc\pytypes\partial.py:20:13 - information: Type of "goo" is "Unknown"
3 errors, 0 warnings, 1 information
```

The issue here is that I want to know about the last 2 errors.  But I _don't_ want to know about the partially unknown type assignment -it's perfectly fine.

This is coming up with the `pandas` type stubs, because `Series` is generic, and if we have a method that returns a `Series`, sometimes we don't know that it is `Series[bool]` or `Series[str]` or whatever, and we are fine with `Series[Unknown]` .  From the point of view of testing the types that would be in a `.pyi` file, knowing that the type of a method is `Unknown` is useful, but knowing that it is "partially unknown" isn't as useful to us.  

**Describe the solution you'd like**
Two options:

- `reportUnknownVariableType` - if `true`, report both partially and fully unknown types
- `reportFullyUnknownVariableType` - if `true`, report only fully unknown types

**Comment by erictraut** [comment]
The need to differentiate between these two cases sounds very specific to your use case. I'd be reluctant to add a new diagnostic check unless we received signal from other uses that this would be more generally useful.

How are your tests structured? Are you invoking pyright on the file and then looking at the diagnostic output (perhaps in JSON form) to validate that it's consistent with some expected set of diagnostics? If so, you could simply filter on the text to differentiate between unknown and partially unknown.

**Comment by hmc-cs-mdrissi** [comment]
I currently have reportUnknownVariableType disabled as too strict a check with 3rd party libraries. I think it'll be significantly easier to enable reportFullyUnknownVariableType than reportFullyUnknownVariableType. I mostly work with tensor libraries and tensorflow/numpy getting data types/shapes to not be unknown is likely far out.

`ndarray[Unknown, dtype[Any]] ` the unknown here is the shape of the array. That is pretty difficult to type hint today. Pep 646 is a great step 1, but there's like several follow up features needed for most numpy functions to be able to support shape hints. Most numpy functions broadcast on shapes and I don't think that's possible to describe in current type system and probably requires custom type operator. My current view is shape types need a couple years of time to develop before that unknown is worth fixing.

The dtype part is also a challenge, but less bad. Unless we have some feature for handling it though my guess is large number of numpy functions will need like dozen+ overloads to handle dtype conversion rules. So possible to do today, but current system makes it extremely verbose and full of duplication.

**Comment by Dr-Irv** [comment]
> I think it'll be significantly easier to enable reportFullyUnknownVariableType than reportFullyUnknownVariableType.

@hmc-cs-mdrissi  can you clarify the above?

I think you meant

I think it'll be significantly easier to enable reportFullyUnknownVariableType than reportUnknownVariableType.

**Comment by Dr-Irv** [comment]
> @Dr-Irv, would it be sufficient to just filter like Eric suggests?

If you can figure out how to do that in the Github workflows action

**Comment by hmc-cs-mdrissi** [comment]
Yes that clarification is correct. I've also edited to your suggestion.

**Comment by Dr-Irv** [comment]
Another related issue.  If you enable `"reportUnknownMemberType": true`, then you detect two things:
```python
df = pd.DataFrame()
df.badMethod()
```
```python
df = pd.DataFrame()
df.iat()
```
I'd like to separate those 2 cases, reporting the first, but not the second.

With the current pandas types, you want to know if you are calling an invalid method (i.e., it's not defined in the `pyi` file) like `badMethod`, but differentiate the case of the method not having a full return type (as in `DataFrame.iat()` currently).  So for `reportUnknownMemberType`, you want to say "tell me when the method is unknown" versus "tell me when the method itself has an unknown type"

**Comment by erictraut** [comment]
We already break out `Unknown` types into 5 separate diagnostic checks:
* reportUnknownParameterType
* reportUnknownArgumentType
* reportUnknownLambdaType
* reportUnknownVariableType
* reportUnknownMemberType

What you're proposing here would require us to double that to 10. I really want to avoid that. Please explore alternative solutions here.

**Comment by Dr-Irv** [comment]
> What you're proposing here would require us to double that to 10. I really want to avoid that. Please explore alternative solutions here.

Maybe the solution is to use the same diagnostic checks, but allow granularity within each check.  So instead of "true", "false", you could have levels defined that are appropriate for each.  See my other example https://github.com/microsoft/pyright/issues/3066#issuecomment-1043556207 with respect to `reportUnknownMemberType`

We need finer granularity on these options to make adding type stubs to existing libraries an easier process that includes automated testing of the stubs.

**Comment by erictraut** [comment]
> I'd like to separate those 2 cases, reporting the first, but not the second.

Normally an access to an undeclared method would result in a `reportGeneralTypeIssues` diagnostic, as in the following:
```python
class Foo: ...
f = Foo()

f.badMethod()
# reportGeneralTypeIssues
# Cannot access member "badMethod" for type "Foo"
#   Member "badMethod" is unknown
```

It works differently with `DataFrame` because it derives from `NDFrame` which defines a `__getattr__` method. This method is invoked for all undeclared member accesses, effectively turning them all into an `Unknown` type.

**Comment by erictraut** [comment]
Nice!
